### ISSUE TITLE
Inconsistent results with/without weight of 0

### ISSUE THREAD

**Issue by gopricy** [crillab/gophersat#18]
Let's say we have 4 vars, [1 2 3 4], and their weights are [1 2 0 1] accordingly.

From my instinct, we can use either 
solver.GtEq([1, 2, 3, 4], [1, 2, 0, 1], 3)
or
solver.GtEq([1, 2, 4], [1, 2, 1], 3)
to describe the condition of L1 + 2L2 + L4 >= 3

But it turns out the above two representation are not equal.
My failure case is as following, you can also try it in Golang Playground https://play.golang.org/p/ddnTIBx3eZO:

``` go
package main

import (
	"fmt"

	"github.com/crillab/gophersat/solver"
)

func main() {
	var lits1, lits2, lits3 []int
	var weights3 []int

	solveAll := func(){
		fmt.Printf("Lits1: %v, Lits2: %v, Lits3: %v, Weights3: %v\n", lits1, lits2, lits3, weights3)
		atMost3 := solver.AtMost(lits1, 3)
		atLeast2 := solver.AtLeast(lits2, 2)
		pbAtLeast3 := solver.GtEq(lits3, weights3, 3)

		pb := solver.ParsePBConstrs(append([]solver.PBConstr{}, atMost3, atLeast2, pbAtLeast3))

		s := solver.New(pb)
		fmt.Println("All solutions", s.CountModels())
	}

	// Case 1
	fmt.Println("Case 1: with weight of 0")
	lits1 = []int{1, 2, 3, 4}
	lits2 = []int{1, 2, 3, 4}
	lits3 = []int{1, 2, 3, 4}

	weights3 = []int{1, 2, 0, 1}
	solveAll()

	// Case 2
	fmt.Println("Case 2: without weight of 0")
	lits1 = []int{1, 2, 3, 4}
	lits2 = []int{1, 2, 3, 4}
	lits3 = []int{1, 2, 4}

	weights3 = []int{1, 2, 1}
	solveAll()
}
```
Let me know if these two representation are not equal by design or this is a bug.
Thanks!

**Comment by frrad** [comment]
This seems like a bug to me too.

**Comment by fdelorme** [comment]
This is definitely a bug, and I'll have a look at it right now. Sorry for the delay, I haven't been notified of this issue by e-mail.

**Comment by gopricy** [comment]
Hi there, thanks for the reply and fix. 
I found another relevant failure case after this fix. So the issue should be rephrased to “Equal PB Problems can lead different num of results”.

Here is another failure case I found
``` go
package main

import (
	"fmt"
	"github.com/crillab/gophersat/solver"
)

func main() {
	pb1 := solver.AtMost([]int{1, 2, 3, 4}, 3)
	pb2 := solver.AtLeast([]int{1, 2, 3, 4}, 2)

	case1 := func(){
		pb3 := solver.GtEq([]int{2, 3, 4}, []int{1, 1, 2}, 3)
		pb := solver.ParsePBConstrs([]solver.PBConstr{pb1, pb2, pb3})
		s := solver.New(pb)
		fmt.Println("Case1", s.CountModels())
	}

	case2 := func(){
		pb3 := solver.GtEq([]int{1, 2, 4}, []int{1, 2, 1}, 3)
		pb := solver.ParsePBConstrs([]solver.PBConstr{pb1, pb2, pb3})
		s := solver.New(pb)
		fmt.Println("Case2", s.CountModels())
	}

	case1()
	case2()
}

// output:
// 4
// 5
```

You can test it out here:
https://play.golang.org/p/lKZ9IlfehRw

@fdelorme can we reopen this issue?

**Comment by fdelorme** [comment]
Yes ! I'll have a look at that.

**Comment by fdelorme** [comment]
The bug is fixed now in the version 1.2.0. It wasn't an issue with PB constraints, but a bug in the `solver.CountModels` functions, caused by an optimization in the core solver.

For instance, as soon as the solver finds the instance is satisfied by binding the variable x1 to false, and the variables x2 and x4 to false, the solver returns as it doesn't have to try to bind the variable x3 (both x3 and not(x3) satisfy the problem). The problem is, `CountModels` counted this as a single model (not(x1) and x2 and x4) instead of 2 (not(x1) and x2 and x3 and x4, or not (x1) and x2 and not(x3) and x4).

As of now the fix doesn't appear in the go playground (probably a caching issue), but I tested it locally and it does now display 5 and 5.

Please let me know if you want me to close the issue.

**Comment by gopricy** [comment]
@fdelorme Thanks for the explanation and it really makes sense. This new release works for me now! I am closing this ticket now.
